relay-server/src/actors/project_buffer.rs

```diff
@@ -290,55 +332,39 @@ impl BufferService {
+                    self.count_mem_envelopes = 0;
+
+                    // spawn the task to enqueue the entire buffer
+                    tokio::spawn(async move {
+                        // A RESERVED lock will be acquired when the first INSERT, UPDATE, or DELETE statement is executed.
+                        // Once we started spooling our in-memory buffer, the dequeue from the disk
+                        // will be blocked till this operations either fails, or commits.
+                        if let Err(err) = sqlx::query("BEGIN").execute(&mut connection).await {
+                            relay_log::error!("failed to start a transaction: {}", LogError(&err));
+                            return;
+                        }
 
-                    // Since we have 3 variables we have to bind, we devide the SQLite limit by 3
-                    // here to prepare the chnunks which will be preparing the batch inserts.
-                    let mut envelopes =
-                        stream::iter(envelopes).chunks(SQLITE_LIMIT_VARIABLE_NUMBER / 3);
-                    while let Some(chunk) = envelopes.next().await {
-                        query_builder.push_values(chunk.into_iter(), |mut b, v| match v.1 {
-                            Ok(envelope_bytes) => {
-                                b.push_bind(v.0.own_key.to_string())
-                                    .push_bind(v.0.sampling_key.to_string())
-                                    .push_bind(envelope_bytes);
-                            }
-                            Err(err) => {
-                                relay_log::error!(
-                                    "failed to serialize the envelope: {}",
-                                    LogError(&err)
-                                )
-                            }
-                        });
-
-                        let query = query_builder.build();
-                        if let Err(err) = query.execute(&db).await {
+                        if let Err(err) = Self::do_spool(&mut connection, buf).await {
                             relay_log::error!(
-                                "failed to buffer envelopes to disk: {}",
+                                "failed to spool memory buffer to the disk: {}",
                                 LogError(&err)
-                            )
+                            );
                         }
-                        // Reset the builder to initial state set by `QueryBuilder::new` function,
-                        // so it can be reused for another chunk.
-                        query_builder.reset();
-                    }
-                });
+
+                        // The SQL command "COMMIT" does not actually commit the changes to disk. It just turns autocommit back on.
+                        // Then, at the conclusion of the command, the regular autocommit logic takes over and causes the actual commit to disk to occur.
```

Comment:
I'm not sure if I understand this correctly; but if I do, this comment is not correct. Since we're using the [WAL journal mode](https://www.sqlite.org/wal.html#how_wal_works), a `COMMIT` will [add a checkpoint](https://www.sqlite.org/wal.html#automatic_checkpoint) (so committing the changes to disk?). Then, indeed, the default committing policy is again in-place.

Reply:
I do not really understand your comment here. 
The comment for this is taken from the docs, just to make sure to clarify what's happening while we call `commit` on the open transaction.

Reply:
Do you mind linking to these docs then? Because I'm misunderstanding something.

Reply:
https://www.sqlite.org/lockingv3.html, see transaction control.

Reply:
Thanks! I misunderstood what checkpointing was, thus my confusion. It makes sense.